torchrl/modules/models/model_based_models.py

```diff
@@ -0,0 +1,316 @@
+        )
+
+
+class RSSMPrior(nn.Module):
+    """The prior network of the RSSM.
+
+    This network takes the previous state and belief and the action and outputs the next state and belief.
+    State is by construction stochastic and belief is deterministic. In the paper this is called the
+    deterministic state and stochastic state respectively. We prefer to call them belief and state for clarity.
+    Furthermore, the paper refers to both prior and posterior as state. We will disntinguish them by calling
+    the prior state and posterior state.
+    The RSSMPrior is a transition model in the sense that it takes the previous state and belief and the action
+    and outputs the next state and belief. It contains a recurrent network that is used to model the temporal
+    dynamics of the state.
+    Other than the state and the belief, the RSSMPrior also outputs the mean and standard deviation of the
+    state distribution. This is useful for the loss function.
+
+    Reference: https://arxiv.org/abs/1811.04551
+
+    Args:
+        hidden_dim (int): Number of hidden units in the linear network. Input size of the recurrent network.
+        rnn_hidden_dim (int): Number of hidden units in the recurrent network. Also size of the belief.
+        state_dim (int): Size of the state.
+        action_spec (TensorSpec): Action spec.
+
+    """
+
+    def __init__(
+        self, hidden_dim=200, rnn_hidden_dim=200, state_dim=30, action_spec=None
+    ):
+        super().__init__()
+
+        # Prior
+        self.rnn = nn.GRUCell(hidden_dim, rnn_hidden_dim)
+        self.action_state_projector = nn.Sequential(nn.LazyLinear(hidden_dim), nn.ELU())
+        self.rnn_to_prior_projector = NormalParamWrapper(
+            nn.Sequential(
+                nn.Linear(hidden_dim, hidden_dim),
+                nn.ELU(),
+                nn.Linear(hidden_dim, 2 * state_dim),
+            ),
+            scale_lb=0,
+            scale_mapping="softplus",
+        )
+
+        self.state_dim = state_dim
+        self.rnn_hidden_dim = rnn_hidden_dim
+
+        if action_spec is None:
+            raise ValueError("action_spec must be provided")
+        self.action_shape = action_spec.shape
+
+    def forward(self, state, rnn_hidden, action):
+        action_state = self.action_state_projector(torch.cat([state, action], dim=-1))
+        rnn_hidden = self.rnn(action_state, rnn_hidden)
+        belief = rnn_hidden
+        prior_mean, prior_std = self.rnn_to_prior_projector(belief)
+        prior_std = prior_std + 0.1
+        prior_state = prior_mean + torch.randn_like(prior_std) * prior_std
+        return prior_mean, prior_std, prior_state, belief
+
+
+class RSSMPosterior(nn.Module):
+    """The posterior network of the RSSM.
+
+    This network takes the belief and the encoded observation and outputs the associated posterior state.
+    It also outputs the mean and standard deviation of the state distribution. This is useful for the loss function.
```

Comment:
`useful` makes it sound like this is optional. I'd rather say `which are inputs to the loss function` or similar.